components/plan-build/bin/hab-plan-build.sh

```diff
@@ -2201,6 +2226,68 @@ do_default_end() {
   return 0
 }
 
+# Internal: Prevent a failure in `do_after_success` from failing the
+# entire job. Execution of this hook is best-effort.
+do_after_success_wrapper() {
+    _do_final_callback_wrapper "do_after_success"
+}
+
+# Internal: Prevent a failure in `do_after_success` from failing the
+# entire job. Execution of this hook is best-effort.
+do_after_failure_wrapper() {
+    _do_final_callback_wrapper "do_after_failure"
+}
+
+# **Internal** Call the `do_after_success` or `do_after_failure`
+# callback functions (if defined), ensuring that errors in *those*
+# functions don't affect the broader build. Failure in a success
+# callback shouldn't fail the build, and the exit code of either
+# shouldn't be able to change the exit code of the build
+# itself. Failure in a callback shouldn't trigger the failure
+# callback.
+_do_final_callback_wrapper() {
+    local callback_function=${1}
+
+    if [[ "$(type -t "${callback_function}")" == "function" ]]; then
+        build_line "'${callback_function}' callback function is defined; executing..."
+
+        # We don't want errors in callback execution to affect the
+        # broader build; execution of these functions is strictly
+        # best-effort, and sugar on top of the build.
+        #
+        # Thus, we need to do a lot of juggling of error handling.
+        #
+        # Kinda gross, but it's Bash.
+
+        # Set this to prevent an error from the hook from triggering
+        # the error handler.
+        set +e
+        # Don't want an error from the success hook to trigger the error hook
+        trap - ERR
+        (
```

Comment:
One consequence of this approach is that any side effects of the `${callback_function}` would be in a sub-shell and would not affect the main process. That is, mutating any variables wouldn't affect the main program. Right now I can't think of a good reason to need this...so we're probably okay here. Is this intended?